### ISSUE TITLE
New rule to forbid access to commonly-confused ambient globals

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#922]
For example, `window.event` ([here](https://github.com/Microsoft/TypeScript/blob/31bbac3c11c82af6cb93fd26ccf1632da6a04353/src/lib/dom.generated.d.ts#L12686) in lib.d.ts) is often referenced in erroneous code like this:

``` ts
function broken(evt: Event) {
    // Meant to do something with `evt` but typed it incorrectly.
    Event.target;  // compiler error
    event.target;  // should be a lint failure
}
```

**Comment by myitcv** [comment]
Also referencing [this thread](https://github.com/Microsoft/TypeScript/issues/1351#issuecomment-160416723) because I'm unclear how such a lint rule could even be defined given the number of such variables.

Perhaps the solution can be approached by using a modified [`lib.dom.d.ts`](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.d.ts) that does _not_ define the globals like `event` but instead relies on the fact that the same objects/functions are available via `window` (which would be one of a well defined, very small number of globals that would include core types). That way, people would be forced to do something like this:

```typescript
function Apple() {
  console.log(event); // error; event no longer defined as a global
  console.log(window.event); // ok
}
```

Thoughts?

Not tested this so this might break some type definitions...?

**Comment by adidahiya** [comment]
@myitcv Is such an approach feasible without [lib.d.ts modularity](https://github.com/Microsoft/TypeScript/issues/494) in TS 2.0?

This issue is fairly similar to #327. Both would almost certainly involve using the type checker (but not necessarily #680 because we could just tell langauge services to consider `lib.d.ts` and the type checker will pick up those global types).

**Comment by myitcv** [comment]
@adidahiya 

> Is such an approach feasible without lib.d.ts modularity in TS 2.0?

It is possible but perhaps not the cleanest solution.

We use custom `lib.core.d.ts` and `lib.dom.d.ts` via the [`--noLib` compiler option](https://github.com/Microsoft/TypeScript/wiki/Compiler-Options). The custom `lib.core.d.ts` and `lib.dom.d.ts` files we use are well defined transformations of the [TypeScript provided versions](https://github.com/Microsoft/TypeScript/tree/master/lib). The transformations we currently apply are:

* All ambient `var`'s in both type definitions are redefined as `const`'s
* Remove [duplicate `URL` definition in `lib.dom.d.ts`](https://github.com/Microsoft/TypeScript/issues/5306)

My plan would be to try a version of `lib.dom.d.ts` (the `var`'s in `lib.core.d.ts` are all core types) that removes all `var` definitions which I would prefer to explicitly access via `window`.

Theoretically, if I published these modified files others would also be able to use the `--noLib` switch, install these type definitions somehow (DefinitelyTyped or similar) and benefit from this modular approach.

**Comment by Patrik-Lundqvist** [comment]
Any progress here? Eslint has the rule [no-restricted-globals](https://eslint.org/docs/rules/no-restricted-globals) which covers this, maybe a similar definition can be used here too?

**Comment by JoshuaKGoldberg** [comment]
Copying discussion from #3824:

What if you declare something in an `externals.d.ts` file, or define your own `lib.dom.d.ts` with either that name or `my-lib.d.ts`?

Or what about a project that uses `namespace`s and has a `const name = "HA!";` somewhere?

Maybe the rule should get the list of files TS is compiling with _(can we do that?)_ along with its compiler options, and add `lib.d.ts`-like files as per `lib` and `--noLib` compiler settings?
Maybe the rule should keep a default list of regular expressions to match names against, like `lib*.d.ts`?

**Comment by Zzzen** [comment]
> What if you declare something in an externals.d.ts file, or define your own lib.dom.d.ts with either that name or my-lib.d.ts?

IMO, there should be no difference between official `lib.dom.d.ts`, 3rd party library declaration and custom declaration.

> Or what about a project that uses namespaces and has a const name = "HA!"; somewhere?

A more concret example would be helpful. Maybe This issue is related. https://github.com/Microsoft/TypeScript/issues/2715

> Maybe the rule should get the list of files TS is compiling with (can we do that?) along with its compiler options, and add lib.d.ts-like files as per lib and --noLib compiler settings?
> Maybe the rule should keep a default list of regular expressions to match names against, like lib*.d.ts?

We are banning globals, no matter where they are declared.

**Comment by cazgp** [comment]
This rule was added: https://palantir.github.io/tslint/rules/no-restricted-globals/
